lib/awful/client.lua

```diff
@@ -1478,8 +1478,10 @@ for _, prop in ipairs { "border_width", "border_color", "opacity" } do
         return self["_"..prop]
     end
     client.object["set_"..prop] = function(self, value)
-        self._private["_user_"..prop] = true
-        self["_"..prop] = value
+        if value ~= nil then
+            self._private["_user_"..prop] = true
+            self["_"..prop] = value
+        end
```

Comment:
So, now it is generally impossible to "unset" something? Uhm.. client shape comes to my mind for something where this is actually used.

(Also, I never understood where this new "set to `nil`" errors come from... what changed that broke stuff that previously worked?)

Reply:
This line is only called for `border_width`, `border_color` and `opacity`. Setting it to `nil` never worked. It is an issue now because my API-backward-compatibility code manages to trigger this. By trying too hard to preserve the API compatibility, I managed to introduce an obscure breakage.

Longer version:

The code to set the border color and width for focused vs. unfocused client was in `rc.lua`. There was also some code paths trying to remove borders for maximized and fullscreen client. Those were mostly fighting each other and "worked" in some case because the signals were connected in a specific way. It was also the only `beautiful` variables which were only used in `rc.lua`.

When I made the permission system and standardized the `request::` signals, I wanted to fix this. Using the permission system allowed the rules to select different border width and color for different clients, etc (well, kind of). It also allowed to add more `beautiful` variables for other states than just `normal` and `focused`. For example `urgent` and `floating`. Finally, it allows to fix the race condition to disable borders for maximized client, which was the point.

To do that change without breaking the API and preserve the behavior when people still had that section in `rc.lua`, I added magic. My magic broke.

Reply:
Thanks for the explanation